include/experimental/__p2642_bits/layout_padded.hpp

```diff
@@ -0,0 +1,742 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+#pragma once
+
+#include <cassert>
+#include "layout_padded_fwd.hpp"
+#include "../__p0009_bits/dynamic_extent.hpp"
+#include "../__p0009_bits/extents.hpp"
+#include "../__p0009_bits/mdspan.hpp"
+#include "../__p0009_bits/layout_left.hpp"
+#include "../__p0009_bits/layout_right.hpp"
+#include "../__p0009_bits/layout_stride.hpp"
+
+namespace std {
+namespace experimental {
+
+namespace detail {
+template<class _T>
+MDSPAN_INLINE_FUNCTION
+constexpr auto
+__find_aligned_offset(_T __alignment, _T __offset)
+{
+  if ( __alignment == 0 )
+    return _T(0);
+  else
+    return ( ( __offset + __alignment - 1 ) / __alignment) * __alignment;
+}
+
+template<class _ExtentsType, size_t _PaddingStride, size_t _ExtentToPadRank>
+MDSPAN_INLINE_FUNCTION
+constexpr size_t
+__get_actual_padding_stride()
+{
+  constexpr auto __rank = _ExtentsType::rank();
+
+  if constexpr (__rank <= size_t(1)) {
+    return _PaddingStride;
+  } else if constexpr (_PaddingStride != dynamic_extent &&
+                       _ExtentsType::static_extent(_ExtentToPadRank) != dynamic_extent) {
+    static_assert((_PaddingStride != 0) || (_ExtentsType::static_extent(_ExtentToPadRank) == 0), "padding stride can be 0 only if extents_type::static_extent(extent-to-pad) is 0");
+    return __find_aligned_offset(_PaddingStride, _ExtentsType::static_extent(_ExtentToPadRank));
+  } else {
+    return dynamic_extent;
+  }
+}
+
+template<size_t _ExtentToSub, class _Extents, size_t _NewExtent, class _Indices>
```

Comment:
Lets rename _ExtentToSub to _ExtentToSubIdx to make it clearer its not the extent value, but the index. In particular since they are all size_t I think this would help.